ext/AdvancedHMCComponentArraysExt.jl

```diff
@@ -0,0 +1,42 @@
+module AdvancedHMCComponentArraysExt
+
+using LinearAlgebra
+
+using AdvancedHMC:
+    AdvancedHMC,
+    Hamiltonian,
+    UnitEuclideanMetric,
+    DiagEuclideanMetric,
+    DenseEuclideanMetric,
+    GaussianKinetic
+using ComponentArrays: ComponentVecOrMat, ComponentVector, ComponentMatrix, getaxes
+
+function AdvancedHMC.∂H∂r(
+    h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::ComponentVecOrMat
+)
+    return copy(r)
+end
+function AdvancedHMC.∂H∂r(
+    h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::ComponentVecOrMat
+)
+    (; M⁻¹) = h.metric
+    @assert getaxes(M⁻¹) == getaxes(r) "Axes of mass matrix and momentum must match"
```

Comment:
Two comments:
- Assertions may be removed by the compiler, IMO it's bad style to use them for actual error checks (I only use them eg to reassert myself - or others - that an `else` branch satisfies certain properties where I know that the assertion should be guaranteed to always be satisfied).
- This requirement seems to be required more generally (and probably be checked internally in LinearAlgebra), so it could just be moved to the generic definition for AbstractVector?

Reply:
> so it could just be moved to the generic definition for AbstractVector?

What do you mean by moving to the generic definition for AbstractVector? We only need to check whether the axes are matching for ComponentArrays.

Reply:
The matching axes are a property that's generally required for most generic matrix multiplications in Julia anyway. It's not really special about ComponentArrays IMO. It's just that it surfaces here since in all other cases that are tested the axes are matching.

Reply:
Actually, it is user's responsibility to input axes-matching ComponentArrays, here for example, axes-matching mass-matrix and momentum, to avoid `Matrix{Float64}*ComponentVector{Float64}=Vector{Float64}` issues. The extension here is to help users check if the axes match instead of allowing the above behaviour.

Reply:
Sure, but why not help users more generally, regardless of whether they use ComponentArrays, regular arrays, or other custom array types?

Reply:
Oh, I see, I will add the axes checks for general arrays as well